### PR TITLE
Refacto/melee phase

### LINKED ISSUE #39: [4.2.4] Modifier le code actuel pour qu'il donne le bon format de prophecy
as:
Non pertinent

i want:
Modifier le code présent actuellement dans la librairie de mathématique pour donner des prophecy qui correspondent au nouveau format de prophecy

description:
Nous avons demandé à des membres de la communauté de IXeme age quelles informations très précisément ils voulaient voir dans les Prophecy produitent par notre code.
Nous allons donc faire en sorte de produire des Prophecy contenant toutes les infos demandées par nos utilisateurs.

definition of done:
Le code présent sur la bibliothèque de mathématique produira dorénavant des Prophecy avec le format demandé par les utilisateurs

workload: 2

### LINKED ISSUE #41: [4.2.3] Modifier le code actuel pour obtenir non pas une mais plusieurs prophecy
as:
Non pertinent

i want:
Le code actuel doit donner toutes les prophecy les plus relevantes et pas juste la plus probable

description:
La librairie de mathématique ne donnera plus une mais plusieurs prophecy en fonction du nombre de fins possible d'un combat

definition of done:
D'après les membres de la communauté de IXeme age, notre application ne devrais pas juste donner la prophecy qui a le plus de chance de se réaliser, mais les résultats de combats les plus probables. Le but est pour lui de savoir quel est le plus grand risque qu'il prends et jusqu'à quel point il peut être gagnant.
Il nous faudra donc modifier le code pour avoir non plus une Prophecy mais une liste de Prophecy a forte probabilités de se réaliser

workload: 4

### LINKED ISSUE #44: [4.5.1] Phase de melee
as:
Non pertinent

i want:
un code dans la librairie de math pour calculer les résultats de la phase de melee

description:
la librairie de math sera maintenant capable de calculer les résultats après une phase de melee

definition of done:
La librairie de math a des fonctions lui permettant de calculer les prophecy en prenant en compte la phase de melee dans un combat 1 unité contre 1 unité. Des tests unitaires seront présent pour ces fonctions

workload: 4

### PR DESCRIPTION
**Pull request by Jordan-B1**
Closes #44 
Closes #41 
Closes #39

### CODE REVIEW

**Review by TheVengeur — APPROVED** [review]

File: src/math_tools/mod.rs
Comment (Pepiloto) [inline src/math_tools/mod.rs:43]: Why not use only one `#`as usual ?

File: src/math_tools/mod.rs
Comment (Pepiloto) [inline src/math_tools/mod.rs:52]: Seems like a Factorial for me maybe rust have a mathematics std call or crate call for that (if it's faster it's interesting else not really)

File: src/math_tools/mod.rs
Comment (Pepiloto) [inline src/math_tools/mod.rs:59]: Can we change the name of the parameter to something clearer ?

File: src/math_tools/mod.rs
Comment (Pepiloto) [inline src/math_tools/mod.rs:61]: Can we change the name of the parameter to something clearer ?

File: src/math_tools/mod.rs
Comment (Pepiloto) [inline src/math_tools/mod.rs:63]: Can we change the name of the parameter to something clearer ?

File: src/math_tools/mod.rs
Comment (Pepiloto) [inline src/math_tools/mod.rs:65]: Why not use only one `#` as usual ?

File: src/prediction/mod.rs
Comment (Pepiloto) [inline src/prediction/mod.rs:9]: Why change the format ?
(last time because I'm too tired to write it again)

File: src/prediction/mod.rs
Comment (Pepiloto) [inline src/prediction/mod.rs:17]: Can't this be modified and added directly to the Regiment struct ?

File: src/prediction/mod.rs
Comment (Pepiloto) [inline src/prediction/mod.rs:60]: What does the name of the variable mean ?

**Review by Pepiloto — CHANGES_REQUESTED** [review]
Don't forget to run the pre-commit hook (ask me if you forgot how)
Don't forget to run `cargo clippy`too
And maybe there is a lot of unneeded `pub` for function
If a struct have a constructor (new) and getter it shouldn't be `pub`

File: src/fight/computation_tools.rs
Comment (Pepiloto) [inline src/fight/computation_tools.rs:64]: No need to put `pub`here, it's only called in this file.

File: src/fight/computation_tools.rs
Comment (Pepiloto) [inline src/fight/computation_tools.rs:136]: No need to put `pub` here, it's only called in this file.

File: src/fight/mod.rs
Comment (Pepiloto) [inline src/fight/mod.rs:163]: No need to put `pub, this is called only in this file.

File: src/fight/mod.rs
Comment (Pepiloto) [inline src/fight/mod.rs:289]: No need to put `pub`

File: src/fight/mod.rs
Comment (Pepiloto) [inline src/fight/mod.rs:363]: Maybe a better variable name

File: src/fight/mod.rs
Comment (Pepiloto) [inline src/fight/mod.rs:385]: No need to put `pub`

File: src/prediction/mod.rs
Comment (Pepiloto) [inline src/prediction/mod.rs:56]: Is that really useful ?

File: src/fight/computation_tools.rs
Comment (Jordan-B1) [reply src/fight/computation_tools.rs:63]: The thing is that: with just one '#' I find the tip box created really unpleasant because of the fat size of the titles.
Actually I did not really respect any norm about those comments in this PR but I can propose this one 
`/// Compute the value to hit the opponent
///
/// ## Parameters
/// offensive (usize): The offensive Stats of the attacking Regiment
///
/// defense (usize): The defense Stats of the defending Regiment`
You can try it on your IDE and if it suits the rest of the team I'll update all of my functions comments for it.
Never seen the problem with the old norm because you never told us you were creating a comment norm and we only saw it on you refacto PR

File: src/fight/computation_tools.rs
Comment (Jordan-B1) [reply src/fight/computation_tools.rs:66]: We call get_stats twice for each regiment, by storing the return value of get_stat() we will avoid one call per regiment
I think its a loss of efficiency to create two variables to avoid only two calls to simple functions like getters

File: Cargo.toml
Comment (Jordan-B1) [reply Cargo.toml:16]: It's not the kind of things we really pay attentions to so you we probably see this on every file file
For this kind of comment, left a comment on the PR rather than on a file. It will be more simple

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs:130]: same for the comments about the get_stat
it's basically just replace a call function by an assignment

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs:132]: It's not really necessary to repeat the same comment two line after :p

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs:186]: bien vu

Comment (Jordan-B1) [reply src/fight/mod.rs:186]: no I can't, I need some data only available in regiment like the number of cols

File: src/fight/computation_tools.rs
Comment (Pepiloto) [reply src/fight/computation_tools.rs:63]: If you prefer the second one that's fine for me.
Can you modify the rest of the comment for the functions then ? So doc is the same everywhere

File: src/fight/computation_tools.rs
Comment (Pepiloto) [reply src/fight/computation_tools.rs:66]: Yes but if you change the parameter to get stats instead of the whole Model you can change it in `compute_save_probability`too and remove 4 calls to get_stats maybe it's better ?

File: src/fight/mod.rs
Comment (Pepiloto) [reply src/fight/mod.rs:130]: It's cleaner to do so and maybe the function will evolve in the future and call getters more.
And a function call is a jmp operation in assembly and assignement is significantly faster because it's registry stocked

File: src/fight/computation_tools.rs
Comment (Pepiloto) [reply src/fight/computation_tools.rs:66]: And still calling a function takes more times than getting a value from a registry

File: src/fight/computation_tools.rs
Comment (Jordan-B1) [reply src/fight/computation_tools.rs:63]: Yes I agree we never discussed about it, i never even thought about it before the refacto but I think that we should ask to everyone if it suits to them before taking one.
And yes ofc if we all agree on one, I'll change the documentation on all my functions as said in my previous comment :+1:

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs]: some left of the old version, imma fix this

**Review by Pepiloto — CHANGES_REQUESTED** [review]
cargo fmt not run
cargo clippy not run
I think we should change global_values::DEFAULT_DICE to f64 because you convert it every time you call it
Does the tests pass (cargo test) ? (at least the unit tests)

**Review by H4ss — COMMENTED** [review]
Code logic is fine. I'll approve after changes to the documentation requested by @Pepiloto or an explanation on the choices.

File: src/fight/computation_tools.rs
Comment (Jordan-B1) [reply src/fight/computation_tools.rs:28]: I'm waiting that the rest of the group say yes for the format. I wont change the comments and do it again if someone have a better idea

File: src/fight/computation_tools.rs
Comment (Jordan-B1) [reply src/fight/computation_tools.rs:10]: oh no you did not mark it on all comments right ? we still have not decided about the format
at least just put a comment on the PR instead

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs:300]: not yet, those are the function I need to test. That's what I was doing before this PR
I'll be back at them after this comment formating thing

File: src/regiment/mod.rs
Comment (Jordan-B1) [reply src/regiment/mod.rs:115]: waiting for the doc format

File: src/regiment/mod.rs
Comment (Jordan-B1) [reply src/regiment/mod.rs:111]: waiting for the doc format

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs]: for the moment I'll let it public because it may become the main function for the fight module
Even if it does not, without this function as pulbic, the rest of the code is never use so I have warnings of non used functions
btw I swapped its name with the compute_turn function. It's more accurate like that

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs]: Yeah, on my todo list

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs]: I think I'll go for something like that but as you can see, the PR is still in draft, I had many little fixes to do

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs:149]: I have a function that handle the ComputeCase::MEAN because it is not computed on the same way than the two others. At the top of the function I return compute_mean_case() if case == ComputeCase::MEAN
So basically I'll throw some holy water on my computer if case is ComputeCase::MEAN at this point

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs:144]: because I return on the top of the function if match is ComputeCase::MEAN

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs:72]: Yeah on my todo list

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs:113]: Error is impossible as param is an optional parameter passed to the function executed by sigma()
it worth None if no parameter has been passed or Some(x) otherwise.
And I gave a parameter like 3lines just under so it can't be None

File: src/fight/mod.rs
Comment (Pepiloto) [reply src/fight/mod.rs:141]: but here if the match is ComputeCase::MEAN it panic no ? How does it come back to the top of the function ?

Comment (Jordan-B1) [reply src/fight/mod.rs:141]: No I mean, at line 93 there is a if statement that say "Hey, if case == ComputeCase::MEAN" return compute_mean_case()
So basically once you reached line 94, case can't be ComputeCase::MEAN
Maybe you getted it with the next comments but just it case that's it :p

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs]: The cloning part is no more here but the point is that we are going to create 3 Predictions for 2 two units
So we need clones of our units before the fight so that we can report 3 different battle results

File: src/fight/mod.rs
Comment (Pepiloto) [inline src/fight/mod.rs:78]: Can we static type the type of this variable ?

File: src/fight/mod.rs
Comment (Pepiloto) [inline src/fight/mod.rs:125]:
```suggestion
/// (&ComputeCase) case: The scenario from first_unit point of view
```

File: src/fight/mod.rs
Comment (Pepiloto) [inline src/fight/mod.rs:146]: Maybe store the return value of `get_regiment_health_points()` in a variable to avoid another jump

File: src/fight/mod.rs
Comment (Pepiloto) [inline src/fight/mod.rs:175]: It's not the right type

File: src/fight/mod.rs
Comment (Pepiloto) [inline src/fight/mod.rs:176]: image
Why this name ?

File: src/fight/mod.rs
Comment (Pepiloto) [inline src/fight/mod.rs:189]: Why empty line ?

File: src/fight/mod.rs
Comment (Pepiloto) [inline src/fight/mod.rs:212]:
```suggestion
/// prediction::Prediction: The prediction computed according to the specified Compute Case
```

File: src/fight/mod.rs
Comment (Pepiloto) [inline src/fight/mod.rs:229]: Add the type `regiment::Regiment`

File: src/regiment/mod.rs
Comment (Pepiloto) [inline src/regiment/mod.rs:109]: Missing `## Parameters`

File: src/regiment/mod.rs
Comment (Pepiloto) [reply src/regiment/mod.rs:140]: Comment already done

**Review by Pepiloto — CHANGES_REQUESTED** [review]

**Comment by Pepiloto** [comment]
More coverage could be really better because it's about some edge cases (for the lines use the tool I talked about on discord)

File: src/regiment/mod.rs
Comment (Jordan-B1) [reply src/regiment/mod.rs:140]: Nope it does not compile: "expected `usize`, found `()`"
It's not really important anyway, it's just a line in a two lines function

File: src/fight/mod.rs
Comment (Jordan-B1) [reply src/fight/mod.rs:78]: I statically typed max_hit but not compute_proba. According to The Book:
A closure expression produces a closure value with a unique, anonymous type that cannot be written out.

**Comment by Jordan-B1** [comment]
I applied the changes. Just after this PR, I'll create a new branch for boost the coverage to min 80% and update the integration tests.
To have as  less people stuck as possible, close your eyes on the coverage for this PR

File: src/fight/computation_tools.rs
Comment (Pepiloto) [inline src/fight/computation_tools.rs:139]: 2 calls to get_aegis should be reunited in a variable

**Review by Pepiloto — DISMISSED** [review]
Just one comment on my side else it's fine for me, you can wait for the other's approval and rebase.

**Review by H4ss — APPROVED** [review]

File: src/fight/computation_tools.rs
Comment (Jordan-B1) [reply src/fight/computation_tools.rs:139]: noted :+1:

**Pull request merged**
